snapshots/wb/2023-08-14/pip_api.py

```diff
@@ -315,74 +342,186 @@ def get_region_data(povline, ppp_version, versions):
-            results = [task.result() for task in concurrent.futures.as_completed(tasks)]
-            # Concatenate list of dataframes
-            results = pd.concat(results, ignore_index=True)
-
-        return results
+                    if Path(
+                        f"{PARENT_DIR}/pip_region_data/pip_region_all_year_all_povline_{povline}_ppp_{ppp_version}.csv"
+                    ).is_file():
+                        log.info(f"Found! (regions, {povline}, {ppp_version} PPPs)")
+                        continue
+                    else:
+                        executor.submit(get_percentiles_data_region, povline, versions, ppp_version)
 
     # Obtain latest versions of the PIP dataset
     versions = pip_versions()
 
     # Run the main function
-    results = concurrent_function()
-    results_region = concurrent_region_function()
+    concurrent_percentiles_function()
+    concurrent_percentiles_region_function()
+
+    log.info("Now we are concatenating the files")
+
+    # Concatenate country and region files
+    df_country = pd.DataFrame()
+    df_region = pd.DataFrame()
+
+    def get_query_country(povline, ppp_version):
+        # Here I check if the file exists even after the original extraction. If it does, I read it. If not, I start the queries again.
+        file_path_country = f"{PARENT_DIR}/pip_country_data/pip_country_all_year_all_povline_{povline}_welfare_all_rep_all_fillgaps_{FILL_GAPS}_ppp_{ppp_version}.csv"
+        if Path(file_path_country).is_file():
+            log.info(f"File to concatenate: countries, {povline}, {ppp_version} PPPs)")
+            df_query_country = pd.read_csv(file_path_country)
+        else:
+            # Run the main function to get the data
+            log.warning(f"We need to come back to the extraction! countries, {povline}, {ppp_version} PPPs)")
+            get_percentiles_data(povline, versions, ppp_version)
+            df_query_country = pd.read_csv(file_path_country)
+
+        return df_query_country
+
+    def get_query_region(povline, ppp_version):
+        # Here I check if the file exists even after the original extraction. If it does, I read it. If not, I start the queries again.
+        file_path_region = (
+            f"{PARENT_DIR}/pip_region_data/pip_region_all_year_all_povline_{povline}_ppp_{ppp_version}.csv"
+        )
+        if Path(file_path_region).is_file():
+            log.info(f"File to concatenate: regions, {povline}, {ppp_version} PPPs)")
+            df_query_region = pd.read_csv(file_path_region)
+        else:
+            # Run the main function to get the data
+            log.warning(f"We need to come back to the extraction! regions, {povline}, {ppp_version} PPPs)")
+            get_percentiles_data_region(povline, versions, ppp_version)
+            df_query_region = pd.read_csv(file_path_region)
+
+        return df_query_region
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
```

Comment:
Hi @Marigold! I turned this file concatenation into a concurrent process, but I see there's not much difference compared to a normal loop. Is it because the process is quick on each step and there's not much to optimize?

Reply:
The issue here is that you **only run** `get_query_country` and `get_query_region` in parallel. The for-loops are still run synchronously because you're calling `.result()` inside them. You can try changing it to this:

```python
futures = []
with concurrent.futures.ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
    for ppp_version in [2011, 2017]:
       for povline in povlines:
           # start the process, but don't call `.result()` which would wait for its completion
           future_country = executor.submit(get_query_country, povline, ppp_version)
           futures.append(future_country)
           
# now that all futures have been started, wait for them all
dfs = [f.result() for f in futures]
df_country = pd.concat(dfs, ignore_index=True)
```

If CPU is the bottleneck, you could also try ProcessPoolExecutor that parallelizes it with cores and not threads.

Reply:
That was it; now it does it in no time. Thanks!